genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/grpc/GRpcAgentFileStreamingServiceImpl.java

```diff
@@ -0,0 +1,388 @@
+        }
+    }
+
+    private static class ReadyStream {
+        private final StreamObserver<AgentJobFileMessage> requestStreamObserver;
+        private final StreamObserver<ServerJobFileMessage> responseStreamObserver;
+
+        ReadyStream(
+            final StreamObserver<AgentJobFileMessage> requestStreamObserver,
+            final StreamObserver<ServerJobFileMessage> responseStreamObserver
+        ) {
+            this.responseStreamObserver = responseStreamObserver;
+            this.requestStreamObserver = requestStreamObserver;
+        }
+
+        ActiveStream beginFileStreaming(
+            final File file,
+            final long startOffset,
+            final long endOffset
+        ) throws IOException {
+            return new ActiveStream(
+                this.requestStreamObserver,
+                this.responseStreamObserver,
+                file,
+                startOffset,
+                endOffset);
+        }
+
+        void sendCompletion() {
+            this.requestStreamObserver.onCompleted();
+        }
+
+        void sendError(final Throwable throwable) {
+            this.requestStreamObserver.onError(throwable);
+        }
+    }
+
+    private static class ActiveStream {
+        private final StreamObserver<AgentJobFileMessage> requestStreamObserver;
+        private final StreamObserver<ServerJobFileMessage> responseStreamObserver;
+        private final File file;
+        private final long startOffset;
+        private final long endOffset;
+        private long watermark;
+        private final byte[] chunkBuffer = new byte[Math.toIntExact(MAX_CHUNK_SIZE)];
+
+        ActiveStream(
+            final StreamObserver<AgentJobFileMessage> requestStreamObserver,
+            final StreamObserver<ServerJobFileMessage> responseStreamObserver,
+            final File file,
+            final long startOffset,
+            final long endOffset
+        ) throws IOException {
+            this.requestStreamObserver = requestStreamObserver;
+            this.responseStreamObserver = responseStreamObserver;
+            this.file = file;
+            this.startOffset = startOffset;
+            this.endOffset = endOffset;
+            this.watermark = startOffset;
+
+            if (startOffset > endOffset) {
+                throw new IllegalArgumentException("End offset is greater than start offset");
+            }
+
+            this.sendFileChunk();
+        }
+
+        private void sendFileChunk() throws IOException {
+            final long bytesLeftToSend = this.endOffset - this.watermark;
+            log.info("{} bytes left to send for file: {}", bytesLeftToSend, file.getAbsolutePath());
+
+            final long chunkSize = Math.min(GRpcAgentFileStreamingServiceImpl.MAX_CHUNK_SIZE, bytesLeftToSend);
+
+            final ByteString chunk;
+            try (FileInputStream fileInputStream = new FileInputStream(this.file)) {
```

Comment:
I'm harping on this a lot in this review so sorry if it's annoying but there's a lot out there about avoiding FIS directly e.g. [here](https://dzone.com/articles/fileinputstream-fileoutputstream-considered-harmful) and [here](https://github.com/google/guava/issues/318)

Reply:
Very interesting. I was gonna switch to channels, but what they suggest here with `Files` is a one-line change..

Reply:
Moved away from direct `FIS` usage (but didn't switch to channels)